Add service for worktree changes

mtsgrd opened the pull request:
Adds service and wires up basic functionality.

- new ChangeList named components will replace FileList*
- rewritten id selection mechanism
- no "ownership" support yet for committing
- make ReduxResult support generic data type
- create new component for worktree changes

<img width="1107" alt="image" src="https://github.com/user-attachments/assets/a5a1f1b7-5169-4af4-a1ae-7c5909a537d1" />

mtsgrd committed:
Add service for worktree changes

mtsgrd committed:
Render names of changed files

- make ReduxResult support generic data type
- create new component for worktree changes

Caleb-T-Owens commented on apps/desktop/src/components/ReduxResult.svelte:
```diff
@@ -1,19 +1,27 @@
-	const { status, error, children }: Props = $props();
+	// eslint-disable-next-line no-undef
+	const { status, data, error, children, empty }: Props<A> = $props();
+	$effect(() => console.log(data));
```
👀

Caleb-T-Owens commented on apps/desktop/src/components/ReduxResult.svelte:
```diff
@@ -1,19 +1,27 @@
-	type Props = {
+	type Props<A> = {
 		status: QueryStatus;
-		error: unknown;
-		children: Snippet;
+		error?: unknown;
+		data?: A;
+		children: Snippet<[A]>;
+		empty: Snippet;
 	};
```
I'm wondering about this API. Is it going to get tedious always restructuring before passing to your result?

I'm just imagining if I had two or three queries in one component, I'd much rather say
```
const foo = getFoo()
const bar = getBar()
const baz = getBaz()
```

rather than

```
const { data: fooData, error: fooError, status: fooStatus } = getFoo()
const { data: barData, error: barError, status: barStatus } = getBar()
const { data: bazData, error: bazError, status: bazStatus } = getBaz()
```

If we work with the whole response object, it also means we can define functions like `map` and `combine` like we have in the web project

mtsgrd committed:
Fix naming mistake

mtsgrd committed:
Render uncommitted file list

- distinguish old FileList items from new with ChangeList name
- rewritten id selection mechanism
- no "ownership" support yet for committing

Caleb-T-Owens commented on apps/desktop/src/lib/worktree/worktreeTypes.ts:
```diff
@@ -78,12 +53,12 @@ export type DiffHunk = {
 	/** Changes that could be committed. */
-	readonly changes: TreeChange[];
+	changes: TreeChange[];
 	/**
 	 * Changes that were in the index that we can't handle.
 	 * The user can see them and interact with them to clear them out before a commit can be made.
 	 */
-	readonly ignoredChanges: IgnoredChange[];
+	ignoredChanges: IgnoredChange[];
```
Why can't these be readonly?

krlvi commented:
Very neat, liking this. One thought, perhaps not for this PR - can we somehow also render the "ignored changes"? Maybe in a separate category and without the ability to interact with them.
I just dont know how to simulate those for testing since it's only exceptional cases that would have these

krlvi commented:
Another thing - the `worktree.ts` (or `worktreeTypes.ts` now) name is probably a bit out of date, because a lot of the types that were added there are now more generically applicable. TreeChange, UnifiedDiff etc are gonna be applicable outside the worktree stuff, but I didnt know where to put it on the frontend

ndom91 reviewed (approved):
Not a big deal, I can do this in my PR if you want, but the empty purgatory styling has been bonked

![image](https://github.com/user-attachments/assets/90067a1f-52c5-4c43-be78-3037eecfb27a)

Otherwise, LGTM! Everythign seems to work

mtsgrd commented:
@ndom91 thanks for spotting. I removed `svg` from the reset css where it was set to display block, and vertical align middle. Those settings are probably worse than default imho, and now that illustration centers.

@krlvi I moved the types to more appropriate places, and yes let's find a way of displaying the ignored files. Landing this for now.
